### ISSUE TITLE
Randomize the order of DNS servers to prevent reaching the limitations of public resolvers

### ISSUE THREAD

**Issue by serge-r** [Jigsaw-Code/outline-apps#1980]
### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Application

Outline Client

### What are you trying to do? What is your use case?

Hello.

I understand that it might seem odd, but I am conducting research on mobile app issues related to Outline VPN, Apple, and Cloudflare's DNS over TLS.

You can find many details along with logs here: https://community.cloudflare.com/t/1-1-1-1-1-0-0-1-dns-over-tls-limitations/643561
I am still awaiting a response from the Cloudflare Community, but it seems like it might take some time.

For those who are unable to follow the link I provided, here is a TL;DR:

When Outline is connected, MacOS (and iOS too) starts using DNS over TLS with Cloudflare servers, because 1.1.1.1 is listed first.

My theory is that Cloudflare has some per-IP limits for DNS DoT queries, and when these limits are reached, I receive a REFUSED response for any domain.

The problem with MacOS X and iOS (which use the same system resolver service, mDNSResponder) when DoT is enabled, is that this negative response is cached for a period, causing all apps that use the system resolver to fail. As a result, retries and other DNS servers configured in the system do not work. This is a significant problem for mobile users, as mobile apps can generate tons of DNS traffic, and iOS users of our mobile app who use Outline are experiencing issues with the application.

I have an Apple developer certificate, and I was able to build and test the Outline Client with different DNS servers. The problems appear only with 1.1.1.1. It seems that changing the order of servers could help us resolve many issues for mobile users.

Im not sure about Windows and Android users, because I don't have devices for testing, but I think would be better to use same behavior on all platforms.

### Is your feature request related to a problem? Please describe it.

Problem not with the Outline but with Cloudflare, details are here: https://community.cloudflare.com/t/1-1-1-1-1-0-0-1-dns-over-tls-limitations/643561

### Describe the solution you'd like.

Would be better to have random order of DNS servers [here](https://github.com/Jigsaw-Code/outline-apps/blob/f00a94cea736697bf4cf10b2cbdf03a48a43b044/client/src/cordova/apple/OutlineAppleLib/Sources/OutlineTunnel/OutlineTunnel.swift#L82) 

### Describe alternatives you've considered

An alternative way is to move 1.1.1.1 from the first position to another, as I did not reproduce the issue with other public resolvers. But Im not sure about others limitations.

**Comment by cornzzy** [comment]
Could also set from dynamic access links

**Comment by serge-r** [comment]
@cornzzy yeah it a variant too, but anyway I need to redefine default behavior because a lot of clients are using defaults and 1.1.1.1 only as a first resolver

**Comment by RomanIzvozchikov** [comment]
My colleagues and I have the same problem. It is really annoying bug. Investigation results are the same as @serge-r described.

We faced this problem on all our Mac computers. The problem occurs on any application that uses mDNSResponder (standard way in MacOS to resolve DNS name).

Interesting fact: we faced this problem on Safari browser (it uses mDNSResponder), but Chrome works better, because it does not rely on mDNSResponder and uses its own DNS resolving mechanism.

So here we have three problems:

1. Cloudflare has some not documented policy that sometimes refuses DNS requests.
2. mDNSResponder written by Apple does not work as expected and does not use other provided DNS servers as a failover mechanism in case of refusing DNS request by Cloudflare.
3. We cannot change DNS server to another one that will work better than hardcoded one in Outline VPN app.

@serge-r thank you for your work!

**Comment by RomanIzvozchikov** [comment]
@daniellacosse @sbruens I don't know when you are going to resolve https://github.com/Jigsaw-Code/outline-apps/issues/568 issue (six years passed), but this problem is really serious for macOS and iOS users. We struggle a lot because of this.

Let's make 9.9.9.9 primary DNS server (it works great, I tested this a lot). I mean just to change one line in your code to 
`["9.9.9.9", "1.1.1.1", "208.67.222.222", "208.67.220.220"]`

I wrote a custom script to workaround 1.1.1.1 DNS reject issue, but I cannot implement this on each computer of employees at my company. We are using Outline VPN as our primary VPN solution. Making 9.9.9.9 as primary server will be great solution until you resolve https://github.com/Jigsaw-Code/outline-apps/issues/568.

**Comment by RomanIzvozchikov** [comment]
@serge-r how did you solve this problem? It is really really annoying problem.

**Comment by serge-r** [comment]
@RomanIzvozchikov Hello!

Sorry I'm not working anymore on this issue and have no idea about any improvements =(

**Comment by fortuna** [comment]
We should take another look at this, since it could be related to https://github.com/OutlineFoundation/outline-apps/issues/2679

**Comment by fortuna** [comment]
I believe we fixed this in https://github.com/OutlineFoundation/outline-apps/pull/2636

https://github.com/OutlineFoundation/outline-apps/blob/a4f2a07738690a3dec06e12b2a896a8b4635582a/client/go/outline/config/outline_dns_intercept.go#L49-L50

@jyyi1 @ohnorobo did we ever release this? Could that PR be the cause of https://github.com/OutlineFoundation/outline-apps/issues/2679?

Of Note: for Windows/Linux (Electron), `client/electron/go_vpn_tunnel.ts:35` still has:

```ts
  const DNS_RESOLVERS = ['1.1.1.1', '9.9.9.9'];
```
 passed as `-tunDNS 1.1.1.1,9.9.9.9` to tun2socks. 

This doesn't randomize, but this issue is specifically about macOS/iOS, and Electron is Windows/Linux only.

**Comment by fortuna** [comment]
@ohnorobo perhaps the solution is to just disable all other resolvers and stick to Quad9. At least on Apple.

**Comment by fortuna** [comment]
After some investigation, it seems the NetworkExtension has a limit of 256 file descriptors, which is a limit on the number of sockets ([reference](https://forum.vizrt.com/index.php?threads/ios-too-many-open-files-with-little-number-of-sources-receivers.250906/#post-1690488)).

Here's a possible explanation for the issue:

  Root cause: VPN loses connectivity after 5-10 minutes due to file descriptor exhaustion in the       
  iOS/macOS Network Extension process. Each DNS query creates a Shadowsocks UDP socket that stays open   
  for 30 seconds (the SDK's write-idle timeout). With 5-10+ DNS queries/second at normal iOS app usage,
  150-300 concurrent Shadowsocks sockets accumulate, hitting the ~256 FD limit of the NE process. New
  sessions then fail silently → DNS stops → all traffic appears broken.

  Fix 1 — dnsintercept/truncate.go: Lazy base session creation.

  Previously, truncatePacketProxy.NewSession() eagerly created a Shadowsocks UDP socket for every UDP
  session, including DNS-only flows that exclusively use the local truncation path and never send
  anything to the remote server. The base session is now created lazily on the first non-DNS packet.
  DNS-only sessions (which is the common case in ppTrunc mode) no longer allocate any Shadowsocks socket
  at all.

  Fix 2 — dnsintercept/forward.go: Close session immediately after first DNS response.

  Previously, after ppForward forwarded a DNS query and received the response, the Shadowsocks UDP socket
   held open for 30 seconds waiting for more traffic that would never come (DNS is one-shot). The
  forwardPacketRespReceiver now calls sender.Close() immediately after delivering the first response from
   the remote resolver, reducing the socket lifetime from ~30 seconds to ~1 network RTT (~100ms).

I'm sending a fix: https://github.com/OutlineFoundation/outline-apps/pull/2724
